RTP Reception hint tracks

polusto opened the pull request:
I've implemented RTP and RTCP Reception Hint track support. Adding support for RTP server hint tracks and other hint tracks should be easier after these changes. 

In hmhd box avgPDUsize and maxPDUsize could probably be implemented better, now they are calculated in append sample. I tried implementing them at the finish movie, but didn't figure out how to access sample data at that point.

Added one new fourcc 'gsm ', because VLC media player seemed to recognize that one better than 'agsm'.

dwbuiten commented:
Before any review, this branch should be cleaned up:
- Rebased to not have a bunch of merge commits
- Not to have a bunch of intermediate WIP commits
- Not to have a commit which changes something, which is changed again in a later commit (squash your commits)

Thanks!

polusto commented:
Ok, no problem. I'll look into it next week.

dwbuiten commented:
Cool, thanks!

polusto commented:
Done. I had to rewrite the commit history, because it would have been too much work to rebase due to conflicts with earlier code and having to fix those conflicts with every commit.

VFR-maniac commented on lsmash.h:
```diff
@@ -719,6 +719,7 @@ DEFINE_QTFF_CODEC_TYPE( QT_CODEC_TYPE_QDMC_AUDIO,    LSMASH_4CC( 'Q', 'D', 'M',
 DEFINE_QTFF_CODEC_TYPE( QT_CODEC_TYPE_QCLP_AUDIO,    LSMASH_4CC( 'Q', 'c', 'l', 'p' ) );    /* Qualcomm PureVoice */
 DEFINE_QTFF_CODEC_TYPE( QT_CODEC_TYPE_AC_3_AUDIO,    LSMASH_4CC( 'a', 'c', '-', '3' ) );    /* Digital Audio Compression Standard (AC-3, Enhanced AC-3) */
 DEFINE_QTFF_CODEC_TYPE( QT_CODEC_TYPE_AGSM_AUDIO,    LSMASH_4CC( 'a', 'g', 's', 'm' ) );    /* GSM */
+DEFINE_QTFF_CODEC_TYPE( QT_CODEC_TYPE_GSM_AUDIO,     LSMASH_4CC( 'g', 's', 'm', ' ' ) );    /* GSM */
```
Any sample file?

VFR-maniac commented on codecs/description.c:
```diff
@@ -272,6 +272,14 @@ static int isom_initialize_structured_codec_specific_data( lsmash_codec_specific
             break;
+        case LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTP_HINT_COMMON:
+            specific->size = sizeof(lsmash_isom_rtp_reception_hint_t);
+            specific->destruct = lsmash_free;
```
Align by '='.

VFR-maniac commented on codecs/description.c:
```diff
@@ -272,6 +272,14 @@ static int isom_initialize_structured_codec_specific_data( lsmash_codec_specific
+            break;
+        case LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTCP_HINT:
+            specific->size = sizeof(lsmash_isom_rtcp_reception_hint_t);
+            specific->destruct = lsmash_free;
```
Ditto.

VFR-maniac commented on codecs/description.c:
```diff
@@ -1536,6 +1554,26 @@ static int isom_set_isom_dts_audio_description( isom_audio_entry_t *audio, lsmas
+{
+    hint->hinttrackversion = summary->version;
+    hint->highestcompatibleversion = summary->highestcompatibleversion;
+    hint->maxpacketsize = summary->maxpacketsize;
```
Ditto.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+int isom_setup_hint_description(isom_stsd_t *stsd, lsmash_hint_summary_t *summary)
+{
+    lsmash_codec_type_t sample_type = summary->sample_type;
+    // Set up sample description for a hint track
```
Don't use // for pure comments.
In the L-SMASH project, // is only used for disabling codes.
Use /* comments */ instead.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+{
+    lsmash_codec_type_t sample_type = summary->sample_type;
+    // Set up sample description for a hint track
+    if (!stsd || !stsd->file || !summary)
```
In L-SMASH conding style, no spaces between statement and parenthesis, and one or more spaces between parentheses and expression, i.e. use if( expression ) style here.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+    // Set up sample description for a hint track
+    if (!stsd || !stsd->file || !summary)
+        return LSMASH_ERR_NAMELESS;
+    int err = isom_check_valid_summary((lsmash_summary_t *)summary);
```
Why no error check here?
Also spaces.

VFR-maniac commented on codecs/description.c:
```diff
@@ -824,6 +838,10 @@ static int isom_check_valid_summary( lsmash_summary_t *summary )
+    else if (lsmash_check_codec_type_identical(sample_type, ISOM_CODEC_TYPE_RRTP_HINT))
+        required_data_type = LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTP_HINT_COMMON;
+    else if (lsmash_check_codec_type_identical(sample_type, ISOM_CODEC_TYPE_RTCP_HINT))
+        required_data_type = LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTCP_HINT;
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -1536,6 +1554,26 @@ static int isom_set_isom_dts_audio_description( isom_audio_entry_t *audio, lsmas
     return 0;
 }
 
+static int isom_set_hint_summary(isom_hint_entry_t *hint, lsmash_hint_summary_t *summary)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -1536,6 +1554,26 @@ static int isom_set_isom_dts_audio_description( isom_audio_entry_t *audio, lsmas
+
+static int isom_set_rtp_additional_reception_description(isom_hint_entry_t *hint, lsmash_hint_summary_t *summary)
+{
+    lsmash_isom_rtp_reception_hint_t *additionaldata = lsmash_malloc(sizeof(lsmash_isom_rtp_reception_hint_t));
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+            if (specific->type == LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTP_HINT_COMMON)
+            {
+                err = isom_set_hint_summary(hint, summary);
+                err = isom_set_rtp_additional_reception_description(hint, summary);
```
Hmmm... I know the 'err' is always set to 0 here, but it  looks strange. Please add comments about it if no error checks.

Spaces.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+    int err = isom_check_valid_summary((lsmash_summary_t *)summary);
+
+    isom_hint_entry_t *hint = isom_add_hint_description(stsd, sample_type);
+    if (!hint)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+    lsmash_codec_type_t hint_type = hint->type;
+
+
+    for (lsmash_entry_t *entry = summary->opaque->list.head; entry; entry = entry->next)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+    {
+        lsmash_codec_specific_t *specific = (lsmash_codec_specific_t *)entry->data;
+
+        if (!specific)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+
+        if (lsmash_check_codec_type_identical(hint_type, ISOM_CODEC_TYPE_RRTP_HINT))
+        {
+            if (specific->type == LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTP_HINT_COMMON)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                rtp_param = (lsmash_isom_rtp_reception_hint_t *)specific->data.structured;
+
+                //mandatory field
+                if (rtp_param->timescale != 0)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                err = isom_set_hint_summary(hint, summary);
+                err = isom_set_rtp_additional_reception_description(hint, summary);
+
+                lsmash_isom_rtp_reception_hint_t* rtp_param;
```
I prefer to type *var; rather than type* var; in C.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                if (rtp_param->timescale != 0)
+                {
+                    isom_tims_t *tims = isom_add_tims(hint);
+                tims->timescale = rtp_param->timescale;
```
irregular indent.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                else
+                {
+                    return LSMASH_ERR_INVALID_DATA;
+                }
```
We omit braces if single statement in block.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                }
+
+                isom_tsro_t *tsro = isom_add_tsro(hint);
+                tsro->offset = rtp_param->time_offset;
```
error check

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                isom_tsro_t *tsro = isom_add_tsro(hint);
+                tsro->offset = rtp_param->time_offset;
+
+                isom_tssy_t *tssy = isom_add_tssy(hint);
```
error check

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+        }
+        else if (lsmash_check_codec_type_identical(hint_type, ISOM_CODEC_TYPE_RTCP_HINT))
+        {
+            if (specific->type == LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTCP_HINT)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+        {
+            if (specific->type == LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTCP_HINT)
+            {
+                lsmash_isom_rtcp_reception_hint_t* rtcp_param;
```
lsmash_isom_rtcp_reception_hint_t *rtcp_param;

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+            {
+                lsmash_isom_rtcp_reception_hint_t* rtcp_param;
+                rtcp_param = (lsmash_isom_rtcp_reception_hint_t *)specific->data.structured;
+                err = isom_set_hint_summary(hint, summary);
```
Please leave comment if no error check.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                err = isom_set_hint_summary(hint, summary);
+                isom_trak_t *trak = (isom_trak_t *)hint->parent->parent->parent->parent->parent;
+                isom_tref_t *tref = isom_add_tref(trak);
+                isom_tref_type_t* tref_type = isom_add_track_reference_type(tref, ISOM_TREF_TYPE_CDSC);
```
null check please

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                isom_trak_t *trak = (isom_trak_t *)hint->parent->parent->parent->parent->parent;
+                isom_tref_t *tref = isom_add_tref(trak);
+                isom_tref_type_t* tref_type = isom_add_track_reference_type(tref, ISOM_TREF_TYPE_CDSC);
+                tref_type->track_ID = malloc(sizeof(uint32_t));
```
spaces and null check

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+    if (err < 0)
+        goto fail;
+
+    return 0;
```
If you check err < 0 for each assigment, I think you need not have if( err < 0 ) goto fail; and return 0; here.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2245,6 +2368,8 @@ int isom_setup_sample_description( isom_stsd_t *stsd, lsmash_media_type media_ty
     else if( media_type == ISOM_MEDIA_HANDLER_TYPE_TEXT_TRACK )
         return isom_setup_text_description( stsd, (lsmash_summary_t *)summary );
+    else if( media_type == ISOM_MEDIA_HANDLER_TYPE_HINT_TRACK)
+        return isom_setup_hint_description(stsd, (lsmash_hint_summary_t *)summary);
```
Spaces

VFR-maniac commented on codecs/description.c:
```diff
@@ -3041,6 +3166,24 @@ int isom_get_implicit_qt_fixed_comp_audio_sample_quants
     return 1;
 }
 
+int hint_update_bitrate(isom_stbl_t *stbl, isom_mdhd_t *mdhd, uint32_t sample_description_index)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.c:
```diff
@@ -3041,6 +3166,24 @@ int isom_get_implicit_qt_fixed_comp_audio_sample_quants
+
+    hmhd->avgPDUsize = hmhd->combinedPDUsize / hmhd->PDUcount;
+
+    err = isom_calculate_bitrate_description(stbl, mdhd, &bufferSizeDB, &maxBitrate, &avgBitrate, sample_description_index);
```
spaces

VFR-maniac commented on codecs/description.c:
```diff
@@ -3074,6 +3217,9 @@ isom_bitrate_updater_t isom_get_bitrate_updater
         RETURN_BITRATE_UPDATER( dts_update_bitrate )
     else if( lsmash_check_codec_type_identical( sample_type, ISOM_CODEC_TYPE_EC_3_AUDIO ) )
         RETURN_BITRATE_UPDATER( eac3_update_bitrate )
+    else if (lsmash_check_codec_type_identical( sample_type, ISOM_CODEC_TYPE_RRTP_HINT )
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on codecs/description.h:
```diff
@@ -43,6 +43,12 @@ int isom_setup_sample_description
+int isom_setup_rtp_hint_description
+(
+    isom_stsd_t      *stsd,
+    lsmash_hint_summary_t *summary
```
Align

VFR-maniac commented on core/box.c:
```diff
@@ -741,7 +742,8 @@ static void isom_remove_audio_description( isom_sample_entry_t *description )
     isom_hint_entry_t *hint = (isom_hint_entry_t *)description;
-    lsmash_free( hint->data );
+    if (hint->additionaldata_length != 0)
+        lsmash_free(hint->additionaldata);
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/box.c:
```diff
@@ -1104,6 +1107,31 @@ static void isom_remove_udta( isom_udta_t *udta )
+        return;
+    lsmash_free(rtp->sdptext);
+    REMOVE_BOX(rtp, isom_hnti_t);
+}
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/box.c:
```diff
@@ -1104,6 +1107,31 @@ static void isom_remove_udta( isom_udta_t *udta )
+        return;
+    lsmash_free(sdp->sdptext);
+    REMOVE_BOX(sdp, isom_hnti_t);
+}
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/box.c:
```diff
@@ -1503,6 +1531,16 @@ isom_audio_entry_t *isom_add_audio_description( isom_stsd_t *stsd, lsmash_codec_
+        return NULL;
+    isom_init_box_common(hint, stsd, sample_type, LSMASH_BOX_PRECEDENCE_HM, isom_remove_hint_description);
+    return isom_add_sample_description_entry(stsd, hint, isom_remove_hint_description) ? NULL : hint;
+}
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/box.c:
```diff
@@ -1669,6 +1710,9 @@ isom_meta_t *isom_add_meta( void *parent_box )
 DEFINE_SIMPLE_BOX_IN_LIST_ADDER( isom_add_cprt, cprt, udta, ISOM_BOX_TYPE_CPRT, LSMASH_BOX_PRECEDENCE_ISOM_CPRT )
+DEFINE_SIMPLE_BOX_ADDER(isom_add_hnti, hnti, udta, ISOM_BOX_TYPE_HNTI, LSMASH_BOX_PRECEDENCE_ISOM_HNTI)
+DEFINE_SIMPLE_BOX_ADDER(isom_add_rtp, rtp, hnti, ISOM_BOX_TYPE_RTP, LSMASH_BOX_PRECEDENCE_ISOM_RTP)
+DEFINE_SIMPLE_BOX_ADDER(isom_add_sdp, sdp, hnti, ISOM_BOX_TYPE_SDP, LSMASH_BOX_PRECEDENCE_ISOM_SDP)
```
Spaces!

VFR-maniac commented on core/box.h:
```diff
@@ -368,6 +368,11 @@ typedef struct
     uint32_t avgbitrate;        /* the average rate in bits/second over the entire presentation */
     uint32_t reserved;
+
+    // temp variables for calculating avgPDUsize
```
Use /* */ instead

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
+    else if (lsmash_check_codec_type_identical(sample_entry->type, ISOM_CODEC_TYPE_RRTP_HINT))
+    {
+        //calculate PDU statistics for hmhd box. 
+        // It requires accessing sample data to get the number of packets per sample.
```
Use /* */ instead for comments.

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
         lsmash_delete_sample( sample );
         return 0;
     }
+    else if (lsmash_check_codec_type_identical(sample_entry->type, ISOM_CODEC_TYPE_RRTP_HINT))
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+(
+lsmash_root_t *root,
+uint32_t       track_ID,
+char          *sdptext
```
Please indents!

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
+
+        // get packetcount in this sample
+        uint16_t packetcount = 0;
+        memcpy(&packetcount, sample->data, sizeof(uint16_t));
```
spaces

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
+        hmhd->combinedPDUsize += sample->length - packetcount*sizeof(rtpsamplecontructor) - 2 * sizeof(uint16_t);
+        hmhd->PDUcount += packetcount;
+
+        for (unsigned int i = 0; i < packetcount; ++i)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
+            memcpy(&cons, data, sizeof(rtpsamplecontructor));
+
+            // constructor type
+            if (cons.type == 2)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948
Also use /* */.

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
+            } // TODO: other constructor types
+        }
+    }
+    else if (lsmash_check_codec_type_identical(sample_entry->type, ISOM_CODEC_TYPE_RTCP_HINT))
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,57 @@ int isom_append_sample_by_type
+        isom_trak_t *track_box = (isom_trak_t*)track;
+        isom_hmhd_t *hmhd = track_box->mdia->minf->hmhd;
+        uint16_t packetcount = 0;
+        memcpy(&packetcount, sample->data, sizeof(uint16_t));
```
spaces

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+        return LSMASH_ERR_FUNCTION_PARAM;
+    lsmash_file_t *file = root->file;
+    if (!file->moov
+        || !file->isom_compatible)
```
weird indentation

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+char          *sdptext
+)
+{
+    if (isom_check_initializer_present(root) < 0 || !sdptext)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+        || !file->isom_compatible)
+        return LSMASH_ERR_NAMELESS;
+    isom_udta_t *udta;
+    if (track_ID)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+    if (track_ID)
+    {
+        isom_trak_t *trak = isom_get_trak(file, track_ID);
+        if (!trak || (!trak->udta && !isom_add_udta(trak)))
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+    }
+    else
+    {
+        if (!file->moov->udta && !isom_add_udta(file->moov))
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+            return LSMASH_ERR_NAMELESS;
+        udta = file->moov->udta;
+    }
+    assert(udta);
```
spaces

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+    hnti = udta->hnti;
+
+    // if track ID is given, text is meant for track hnti box, 
+    // otherwise it is meant for movie hnti box
```
Use /* */ instead.

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+            return LSMASH_ERR_NAMELESS;
+
+        isom_rtp_t* rtp = hnti->rtp;
+        rtp->descriptionformat = '\x73\x64\x70'; //'sdp '
```
Wait 'sdp ' is not '\x73\x64\x70\x20'? how do you handle 4th byte?

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+        isom_rtp_t* rtp = hnti->rtp;
+        rtp->descriptionformat = '\x73\x64\x70'; //'sdp '
+        rtp->sdp_length = strlen(sdptext); /* leaves \0 out*/
+        rtp->sdptext = lsmash_memdup(sdptext, rtp->sdp_length);
```
spaces

VFR-maniac commented on core/summary.c:
```diff
@@ -234,6 +237,15 @@ int lsmash_compare_summary( lsmash_summary_t *a, lsmash_summary_t *b )
+        lsmash_hint_summary_t *out_hint = (lsmash_hint_summary_t *)b;
+        if (in_hint->version != out_hint->version
+            || in_hint->highestcompatibleversion != out_hint->highestcompatibleversion
+            || in_hint->maxpacketsize != out_hint->maxpacketsize)
```
alignments!

VFR-maniac commented on core/write.c:
```diff
@@ -469,6 +469,32 @@ static int isom_write_btrt( lsmash_bs_t *bs, isom_box_t *box )
+    lsmash_bs_put_byte(bs, tssy->reserved_sync);
+
+    return 0;
+}
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/write.c:
```diff
@@ -598,20 +624,26 @@ static int isom_write_audio_description( lsmash_bs_t *bs, isom_box_t *box )
-    isom_hint_entry_t *data = (isom_hint_entry_t *)entry->data;
-    if( !data )
+    isom_hint_entry_t *data = (isom_hint_entry_t *)box;
+    if (!data)
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on core/write.c:
```diff
@@ -598,20 +624,26 @@ static int isom_write_audio_description( lsmash_bs_t *bs, isom_box_t *box )
+
+    lsmash_bs_put_be16(bs, data->hinttrackversion);
+    lsmash_bs_put_be16(bs, data->highestcompatibleversion);
+    lsmash_bs_put_be32(bs, data->maxpacketsize);
```
spaces!

VFR-maniac commented on core/write.c:
```diff
@@ -598,20 +624,26 @@ static int isom_write_audio_description( lsmash_bs_t *bs, isom_box_t *box )
+    lsmash_bs_put_be16(bs, data->highestcompatibleversion);
+    lsmash_bs_put_be32(bs, data->maxpacketsize);
+
+    // additional data boxes are written separately
```
/* */

VFR-maniac commented on core/write.c:
```diff
@@ -1079,6 +1111,28 @@ static int isom_write_udta( lsmash_bs_t *bs, isom_box_t *box )
+    isom_bs_put_box_common(bs, sdp);
+    lsmash_bs_put_bytes(bs, sdp->sdp_length, sdp->sdptext);
+    return 0;
+}
```
ref: https://github.com/l-smash/l-smash/pull/55/files#r43904948

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+(
+lsmash_root_t *root,
+uint32_t       track_ID,
+char          *sdptext
```
Please indents!

VFR-maniac commented on lsmash.h:
```diff
@@ -1413,6 +1418,18 @@ int lsmash_convert_clap_into_crop
+    LSMASH_BASE_SUMMARY
+    uint16_t version;
+    uint16_t highestcompatibleversion;
+    uint32_t maxpacketsize; // maximum size of one packet
```
Use /* */.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+{
+    int32_t relative_time;
+
+    // from RTP Header
```
/**/

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+
+typedef struct
+{
+    uint8_t type; // this is type 2 constructor
```
/**/

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+    ISOM_SYNC_ANY = 0,
+    ISOM_SYNC_RECIEVED_RTP = 1,
+    ISOM_SYNC_OTHER_TRACKS = 2,
+    ISOM_SYNC_RESERVED = 3,
```
Align by '='.

VFR-maniac commented on core/box.h:
```diff
@@ -453,6 +458,28 @@ typedef struct
+{
+    /* This box is in RTP reception hint track sample description */
+    ISOM_BASEBOX_COMMON;
+    uint8_t reserved_sync;
```
unsigned reserved       : 6;
unsigned timestamp_sync : 2;
Please write original forms described in the spec as much as possible if not exposed as library I/F.
This helps search of technical terms in documents.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+    uint16_t seq_number;
+    uint16_t reserved_flags;
+    uint16_t entrycount;
+}RTPpacket;
```
A space before RTPpacket.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+    uint32_t sampleoffset;
+    uint16_t bytesperblock;
+    uint16_t samplesperblock;
+}rtpsamplecontructor;
```
A space before rtpsamplecontructor.

VFR-maniac commented on core/box.h:
```diff
@@ -2679,6 +2752,9 @@ isom_stco_t *isom_add_stco( isom_stbl_t *stbl );
 isom_cprt_t *isom_add_cprt( isom_udta_t *udta );
+isom_hnti_t *isom_add_hnti( isom_udta_t *udta );
+isom_rtp_t  *isom_add_rtp( isom_hnti_t *hnti);
+isom_sdp_t  *isom_add_sdp( isom_hnti_t *hnti);
```
spaces!

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+
+    hint->data_reference_index = summary->data_ref_index;
+
+    // configure the sample description
```
Use /* */

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2276,90 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                lsmash_isom_rtp_reception_hint_t* rtp_param;
+                rtp_param = (lsmash_isom_rtp_reception_hint_t *)specific->data.structured;
+
+                //mandatory field
```
Use /* */

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+
+        isom_sdp_t* sdp = hnti->sdp;
+        sdp->sdp_length = strlen(sdptext);/* leaves \0 out*/
+        sdp->sdptext = lsmash_memdup(sdptext, sdp->sdp_length);
```
null check please.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,63 @@ int lsmash_create_object_descriptor
+
+typedef struct
+{
+	uint32_t rtp_track_id; // which track this RTCP track is referencing
```
odd indentation

VFR-maniac commented on core/box.c:
```diff
@@ -1104,6 +1107,31 @@ static void isom_remove_udta( isom_udta_t *udta )
+        return;
+
+    REMOVE_BOX(hnti, isom_udta_t);
+}
```
where are isom_remove_rtp() and isom_remove_sdp() invoked?

VFR-maniac commented on core/box.c:
```diff
@@ -1104,6 +1107,31 @@ static void isom_remove_udta( isom_udta_t *udta )
+    if (!hnti)
+        return;
+
+    REMOVE_BOX(hnti, isom_udta_t);
```
DEFINE_SIMPLE_BOX_REMOVER( isom_remove_hnti, hnti, isom_udta_t )

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+        if (!isom_add_sdp(hnti))
+            return LSMASH_ERR_NAMELESS;
+
+        isom_sdp_t* sdp = hnti->sdp;
```
isom_sdp_t __*__sdp

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4242,63 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+    }
+    assert(udta);
+
+    isom_hnti_t* hnti;
```
isom_hnti_t *hnti;

VFR-maniac commented:
Deeper reviews may be occurred if I have time.

VFR-maniac commented:
You need bump minor ver. of lsmash.h.

polusto commented:
I don't have a gsm file at hand, but I can probably generate one next week. Did everything mentioned, unless I missed something. Have a nice weekend everyone!

VFR-maniac commented:
Please do not do everything at single commit.
git reset HEAD~1
&
git add -p
are your friends.

VFR-maniac commented:
Still if ( expression ) is there. please remove space between if and (.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,82 @@ int lsmash_create_object_descriptor
+    uint16_t length;
+    uint32_t sampledescriptionindex;
+    uint32_t sampledescriptionoffset;
+} rtpsampledescriptioncontructor;
```
Ditto.

And I think these constructors can be integrated as a union as the spec says.

VFR-maniac replied:
Ditto.

And I think these constructors can be integrated as a union as the spec says.

VFR-maniac replied:
By the integration, for instance, the helper function could be the following.
lsmash_sample_t *lsmash_make_rtp_sample( lsmash_rtp_packet_t packet, lsmash_rtp_constructor_t constructor);
Of course, this is the simplest case. RTPsample could be more complicated structures.

VFR-maniac replied:
rtpsampledescriptioncon*s*tructor

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,82 @@ int lsmash_create_object_descriptor
+    uint16_t reserved_flags; /* reserved(13), extra_flag(1), bframe_flag(1), repeat_flag(1) */
+    uint16_t entrycount;
+
+} RTPpacket;
```
Your I/F is not user-friendly. RTPpacket is not meaningful as an API. There is no necessity as a public API.
If you add this here, please consider adding helper functions to construct RTPsample defined in 9.1.3 Sample Format as public APIs.

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,58 @@ int isom_append_sample_by_type
+        /* get packetcount in this sample */
+        uint16_t packetcount = 0;
+        memcpy( &packetcount, sample->data, sizeof(uint16_t) );
+        uint8_t *data = sample->data + 2 * sizeof(uint16_t) + sizeof(RTPpacket);
```
I think sizeof(RTPpacket) is dangerous since alignments by compilers may add more bytes than what you expected.

VFR-maniac replied:
Basically, you should not use sizeof operator to decide offsets of binary strings provided by users or written to streams. Different compilers may produce different sizes from the same data types. This also applies to sizeof(uint16_t) here.

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,58 @@ int isom_append_sample_by_type
+        uint8_t *data = sample->data + 2 * sizeof(uint16_t) + sizeof(RTPpacket);
+
+        /* calculate only packet headers and packet payloads sizes int PDU size. Later use these two to get avgPDUsize */
+        hmhd->combinedPDUsize += sample->length - packetcount*sizeof(rtpsamplecontructor) - 2 * sizeof(uint16_t);
```
Ditto about sizeof(rtpsamplecontructor).

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3971,58 @@ int isom_append_sample_by_type
+        for( unsigned int i = 0; i < packetcount; ++i )
+        {
+            rtpsamplecontructor cons;
+            memcpy( &cons, data, sizeof(rtpsamplecontructor) );
```
Ditto about sizeof(rtpsamplecontructor).

I realize that all constructors are 16 bytes. so you can define this value as a macro and use this.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,82 @@ int lsmash_create_object_descriptor
+    uint16_t seq_number;
+    uint16_t reserved_flags; /* reserved(13), extra_flag(1), bframe_flag(1), repeat_flag(1) */
+    uint16_t entrycount;
+
```
An extra line :P

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2278,117 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+        }
+        else if ( lsmash_check_codec_type_identical( hint_type, ISOM_CODEC_TYPE_RTCP_HINT ) )
+        {
+            if ( specific->type == LSMASH_CODEC_SPECIFIC_DATA_TYPE_ISOM_RTCP_HINT )
```
I do not like the I/F which conveys track reference info via sample description layer.
There is big wall between track/presentation layer and media/sample layer as the concepts of ISOBMFF.
Please reconsider this.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2278,117 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                isom_tsro_t *tsro = isom_add_tsro( hint );  /* not necessary if time_offset = 0, but add anyway */
+                isom_tssy_t *tssy = isom_add_tssy( hint );
+
+                if ( rtp_param->timescale == 0 )
```
This check can be placed a bit earlier, can't this?
It is not preferable that this check comes between additions of additional boxes and null checks of these boxes.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2278,117 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                if (!tref_type->track_ID)
+                    err = LSMASH_ERR_MEMORY_ALLOC;
+
+                *(tref_type->track_ID) = rtcp_param->rtp_track_id;
```
NULL dereference could cause here.

VFR-maniac commented on lsmash.h:
```diff
@@ -4086,6 +4103,82 @@ int lsmash_create_object_descriptor
+    uint32_t sampleoffset;
+    uint16_t bytesperblock;
+    uint16_t samplesperblock;
+} rtpsamplecontructor;
```
rtpsamplecon*s*tructor

VFR-maniac commented on core/box.c:
```diff
@@ -741,7 +742,8 @@ static void isom_remove_audio_description( isom_sample_entry_t *description )
     isom_hint_entry_t *hint = (isom_hint_entry_t *)description;
-    lsmash_free( hint->data );
+    if ( hint->additionaldata_length != 0 )
+        lsmash_free( hint->additionaldata );
```
You can just write lsmash_free( hint->additionaldata ); here.
The lsmash_free() function does nothing when passing nullptr.

VFR-maniac commented on core/box.c:
```diff
@@ -1104,10 +1107,28 @@ static void isom_remove_udta( isom_udta_t *udta )
     }
 }
 
+
```
an extra line.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2278,117 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+                lsmash_isom_rtp_reception_hint_t *rtp_param;
+                rtp_param = ( lsmash_isom_rtp_reception_hint_t *)specific->data.structured;
+
+
```
Why are there three blank lines here?

VFR-maniac commented on core/box.h:
```diff
@@ -368,6 +368,11 @@ typedef struct
+    /* temp variables for calculating avgPDUsize */
+    uint64_t combinedPDUsize;
+    uint64_t PDUcount;
+
```
an extra blank line

VFR-maniac commented on core/summary.c:
```diff
@@ -334,6 +347,8 @@ lsmash_codec_support_flag lsmash_check_codec_support( lsmash_codec_type_t sample
         ADD_CODEC_SUPPORT_TABLE_ELEMENT( QT_CODEC_TYPE_V410_VIDEO,      LSMASH_CODEC_SUPPORT_FLAG_REMUX );
         ADD_CODEC_SUPPORT_TABLE_ELEMENT( QT_CODEC_TYPE_YUV2_VIDEO,      LSMASH_CODEC_SUPPORT_FLAG_REMUX );
+        ADD_CODEC_SUPPORT_TABLE_ELEMENT( ISOM_CODEC_TYPE_RRTP_HINT,     LSMASH_CODEC_SUPPORT_FLAG_REMUX );
+        ADD_CODEC_SUPPORT_TABLE_ELEMENT( ISOM_CODEC_TYPE_RTCP_HINT,     LSMASH_CODEC_SUPPORT_FLAG_REMUX );
```
I don't think you add supports of hint track demuxers.

polusto commented:
Thanks. Unfortunately, I may not have time to fix these before December.

polusto commented:
Hi again. Uploaded a a new version of pull request. I implemented the reference boxes as a separate function. I decided to leave out the gsm commit. I also went through the code correcting style the best I could and also addressed other issues mentioned.

Happy new year.

polusto commented:
Any possibility for a review?

VFR-maniac commented:
Sorry, recently I'm really busy, I'll look into when I have time.

VFR-maniac commented on codecs/description.c:
```diff
@@ -1536,6 +1547,16 @@ static int isom_set_isom_dts_audio_description( isom_audio_entry_t *audio, lsmas
 
+static int isom_set_hint_summary( isom_hint_entry_t *hint, lsmash_hint_summary_t *summary )
+{
+    hint->data        = lsmash_malloc( sizeof(isom_rtp_hint_entry_t) );
```
Check the allocation if successful or not here and return LSMASH_ERR_MEMORY_ALLOC if failure.

VFR-maniac commented on codecs/description.c:
```diff
@@ -2237,6 +2258,66 @@ static int isom_setup_text_description( isom_stsd_t *stsd, lsmash_summary_t *sum
+        }
+    }
+    else if( lsmash_check_codec_type_identical(hint_type, ISOM_CODEC_TYPE_RTCP_HINT ) )
+        isom_set_hint_summary( hint, summary );
```
return isom_set_hint_summary( hint, summary );

VFR-maniac commented on codecs/description.c:
```diff
@@ -1536,6 +1547,16 @@ static int isom_set_isom_dts_audio_description( isom_audio_entry_t *audio, lsmas
+    hint->data_length = sizeof(isom_rtp_hint_entry_t);
+    LSMASH_SET_BE16( hint->data,     summary->version );
+    LSMASH_SET_BE16( hint->data + 2, summary->highestcompatibleversion );
+    LSMASH_SET_BE32( hint->data + 4, summary->maxpacketsize );
```
wait, what. This is dangerous. Do not treat as binary string if allocated by sizeof(struct_type).
The sizeof(struct_type) depends on compiler implementation, so don't assume fixed offsets for each field.

VFR-maniac replied:
If you do this, use hint->data  = lsmash_malloc( 8 ); /* 8 = (16+16+32)/8 */.
But I don't recommend that since it's inconvenient to handle internally. BInary string format should be used only for exporting data to outside of L-SMASH library and reading it. Otherwise opaque pointer should be used. That is why there are LSMASH_CODEC_SPECIFIC_FORMAT_STRUCTURED and LSMASH_CODEC_SPECIFIC_FORMAT_UNSTRUCTURED.

VFR-maniac replied:
I think the `data` field should be deleted from ISOM_HINT_SAMPLE_ENTRY or be changed into an opaque pointer since almost all derived hint sample entries do extend SampleEntry instead of HintSampleEntry.

polusto replied:
I did not understand this part. I know there are different lengths fo hint sample entries, but how do you want them handled inside L-SMASH?

VFR-maniac commented on codecs/description.c:
```diff
@@ -3041,6 +3124,21 @@ int isom_get_implicit_qt_fixed_comp_audio_sample_quants
+    uint32_t avgBitrate = 0;
+    isom_hmhd_t *hmhd = ( (isom_mdia_t*)(mdhd->parent) )->minf->hmhd;
+    int err = LSMASH_ERR_NAMELESS;
+    hmhd->avgPDUsize = hmhd->combinedPDUsize / hmhd->PDUcount;
```
I think zero-division could occur here.

VFR-maniac commented on core/isom.c:
```diff
@@ -1294,6 +1291,13 @@ isom_trak_t *isom_track_create( lsmash_file_t *file, lsmash_media_type media_typ
+            trak->mdia->minf->hmhd->PDUcount = 0;
+            break;
+        case ISOM_MEDIA_HANDLER_TYPE_RTCP_RECEPTION_HINT_TRACK :
+            if (!isom_add_nmhd(trak->mdia->minf))
```
spaces!

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4230,122 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+    if( !tref_type )
+        return LSMASH_ERR_NAMELESS;
+
+    tref_type->track_ID = malloc( sizeof(uint32_t) );
```
lsmash_malloc

VFR-maniac commented on core/write.c:
```diff
@@ -1507,6 +1556,8 @@ void isom_set_box_writer( isom_box_t *box )
             box->write = isom_write_visual_description;
         else if( media_type == ISOM_MEDIA_HANDLER_TYPE_AUDIO_TRACK )
             box->write = isom_write_audio_description;
+        else if( media_type == ISOM_MEDIA_HANDLER_TYPE_HINT_TRACK || media_type == ISOM_MEDIA_HANDLER_TYPE_HINT_TRACK )
```
???

VFR-maniac commented on core/isom.c:
```diff
@@ -3966,6 +3979,37 @@ int isom_append_sample_by_type
+        /* get packetcount in this sample */
+        uint16_t packetcount = 0;
+        memcpy( &packetcount, sample->data, sizeof(uint16_t) );
+        uint8_t *data = sample->data + 2 * sizeof(uint16_t) + RTP_PACKET_SIZE;
```
Do not use sizeof operator for an offset from binary string, it's a compiler dependent thing. Define RTP_SAMPLE_PACKET_COUNT_SIZE (= 2) and use it.

VFR-maniac replied:
Or just do as follows.
```
uint16_t packet_count = LSMASH_GET_BE16( sample->data );
uint8_t *data = sample->data + RTP_SAMPLE_HEADER_SIZE + RTP_PACKET_SIZE;
```

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4230,122 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+
+int lsmash_add_track_reference( lsmash_root_t *root, uint32_t track_ID, uint32_t ref_track_ID, isom_track_reference_type type )
+{
+    if( isom_check_initializer_present(root) < 0
```
spaces around `root`.

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4230,122 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+    if( track_ID )
+    {
+        isom_trak_t *trak = isom_get_trak( file, track_ID );
+        if( !trak ||( !trak->udta && !isom_add_udta( trak ) ) )
```
if( !trak || (!trak->udta && !isom_add_udta( trak )) )

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4230,122 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+
+        isom_sdp_t *sdp = hnti->sdp;
+        sdp->sdp_length = strlen( sdptext );/* leaves \0 out*/
+        sdp->sdptext = lsmash_memdup( sdptext, sdp->sdp_length );
```
no null check?

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4230,122 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+        isom_rtp_t *rtp = hnti->rtp;
+        rtp->descriptionformat = '\x73\x64\x70\x20'; /* 'sdp ' */
+        rtp->sdp_length = strlen( sdptext ); /* leaves \0 out */
+        rtp->sdptext = lsmash_memdup( sdptext, rtp->sdp_length );
```
no null check?

VFR-maniac commented on core/isom.c:
```diff
@@ -4186,3 +4230,122 @@ int lsmash_set_copyright( lsmash_root_t *root, uint32_t track_ID, uint16_t ISO_l
+        tref = isom_add_tref( trak );
+
+    if( !tref )
+        return LSMASH_ERR_NAMELESS;
```
you can do as follows.
```
if( !tref && !(tref = isom_add_tref( trak )) )
    return LSMASH_ERR_NAMELESS;
```

polusto commented:
I removed RTCP reception hint track to make reviewing easier. I'm going to make a second pull request for tref box. I tried my best fix all the issues mentioned.

VFR-maniac commented:
I'm sorry for extremely delay of merging: https://github.com/l-smash/l-smash/commit/d2571a1d50b246e97b6ebe3c3e14804a4addad75
